compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1335,7 +1335,10 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         case RefinedType(parent, nme.apply, mt @ MethodTpe(_, formals, restpe))
         if (defn.isNonRefinedFunction(parent) || defn.isErasedFunctionType(parent)) && formals.length == defaultArity =>
           (formals, untpd.DependentTypeTree(syms => restpe.substParams(mt, syms.map(_.termRef))))
-        case SAMType(mt @ MethodTpe(_, formals, restpe)) =>
+        case pt1 @ SAMType(mt @ MethodTpe(_, formals, methResType)) =>
+          val restpe = methResType match
+            case mt: MethodType if !mt.isParamDependent => mt.toFunctionType(isJava = pt1.classSymbol.is(JavaDefined))
```

Comment:
What happens if `mt` _is_ ParamDependent? Can we still get a crash then? 

Maybe we should in that case fall back to the default composition with wildcard types?

Reply:
I tried and I think if you make it param dependent it doesn't qualify as a SAM type?  I guess it'll be good to have a case in, just to prove it doesn't crash the compiler - and that way you can review that I've written the test case correctly.

Reply:
> I tried and I think if you make it param dependent it doesn't qualify as a SAM type?

 I doubt that. Types with ParamDependent method types cannot be SAM types. But I don't see that ParamDependent result types are excluded.